### PR TITLE
 Allow registering custom TemplateContextCollector's

### PR DESCRIPTION
**Pull request by raffaelecarelle**
Fixes https://github.com/twigstan/twigstan/issues/71

### CODE REVIEW

File: src/Config/ConfigBuilder.php [hunk 1]
```diff
@@ -20,6 +21,10 @@ final class ConfigBuilder
     private ?string $phpstanConfigurationFile = null;
     private null | false | string $phpstanMemoryLimit = null;
     private ?string $twigEnvironmentLoader = null;
+    /**
+     * @var list<class-string>
```
Comment (ruudk) [inline src/Config/ConfigBuilder.php:25]:
```suggestion
     * @var list<class-string<TemplateContextCollector>>
```

File: src/Application/PHPStanRunner.php [hunk 2]
```diff
@@ -58,6 +62,14 @@ public function run(
             $parameters['customRulesetUsed'] = true;
         }
 
+        $services = [];
+        foreach ($this->twigContextCollector as $className) {
+            $services[] = [
+                'class' => $className,
+                'tags' => ['phpstan.collector'],
+            ];
+        }
```
Comment (ruudk) [inline src/Application/PHPStanRunner.php:71]: Only do this when `$collectOnly`

File: src/Config/ConfigBuilder.php [hunk 3]
```diff
@@ -20,6 +21,10 @@ final class ConfigBuilder
...
+     */
+    private array $twigContextCollector = [];
```
Comment (ruudk) [inline src/Config/ConfigBuilder.php:27]:
```suggestion
    private array $twigContextCollectors = [];
```

File: src/Config/ConfigBuilder.php [hunk 4]
```diff
@@ -147,6 +152,7 @@ public function create(): TwigStanConfig
             $this->phpExcludes,
             $this->ignoreErrors,
             $this->baselineErrors,
+            $this->twigContextCollector,
```
Comment (ruudk) [inline src/Config/ConfigBuilder.php:155]:
```suggestion
            $this->twigContextCollectors,
```

File: src/Config/ConfigBuilder.php [hunk 5]
```diff
@@ -376,4 +382,19 @@ public function ignoreErrors(IgnoreError ...$ignoreErrors): self
 
         return $this;
     }
+
+    /**
+     * @param class-string ...$classNames
```
Comment (ruudk) [inline src/Config/ConfigBuilder.php:387]:
```suggestion
     * @param class-string<TemplateContextCollector> ...$classNames
```

File: src/Config/ConfigBuilder.php [hunk 6]
```diff
@@ -376,4 +382,19 @@ public function ignoreErrors(IgnoreError ...$ignoreErrors): self
...
+     */
+    public function twigContextCollector(string ...$classNames): self
+    {
+        foreach ($classNames as $className) {
+            if ( ! is_a($className, TemplateContextCollector::class, true)) {
+                throw new RuntimeException(sprintf('Class %s does not implement %s interface.', $className, TemplateContextCollector::class));
+            }
+            $this->twigContextCollector[] = $className;
```
Comment (ruudk) [inline src/Config/ConfigBuilder.php:395]:
```suggestion
            $this->twigContextCollectors[] = $className;
```

File: src/Config/TwigStanConfig.php [hunk 7]
```diff
@@ -16,6 +16,7 @@
      * @param list<string> $phpExcludes
      * @param list<IgnoreError> $ignoreErrors
      * @param list<BaselineError> $baselineErrors
+     * @param list<class-string> $twigContextCollector
```
Comment (ruudk) [inline src/Config/TwigStanConfig.php:19]:
```suggestion
     * @param list<class-string<TemplateContextCollector>> $twigContextCollectors
```

File: src/Config/TwigStanConfig.php [hunk 8]
```diff
@@ -32,6 +33,7 @@ public function __construct(
         public array $phpExcludes,
         public array $ignoreErrors,
         public array $baselineErrors,
+        public array $twigContextCollector,
```
Comment (ruudk) [inline src/Config/TwigStanConfig.php:36]:
```suggestion
        public array $twigContextCollectors,
```

File: src/DependencyInjection/ContainerFactory.php [hunk 9]
```diff
@@ -62,6 +62,7 @@ public function create(): Container
             'phpPaths' => $this->configuration->phpPaths,
             'phpExcludes' => $this->configuration->phpExcludes,
             'ignoreErrors' => $this->configuration->ignoreErrors,
+            'twigContextCollector' => $this->configuration->twigContextCollector,
```
Comment (ruudk) [inline src/DependencyInjection/ContainerFactory.php:65]:
```suggestion
            'twigContextCollectors' => $this->configuration->twigContextCollectors,
```

**Review by ruudk — COMMENTED** [review]
Great! Let's make sure the variable is always plural. And only define the services when `$collectOnly`.

After that, could you add a test case as well?

File: src/Application/PHPStanRunner.php [hunk 10]
```diff
@@ -14,6 +14,9 @@
 
 final readonly class PHPStanRunner
 {
+    /**
+     * @param list<class-string<TemplateContextCollector>> $twigContextCollectors
```
Comment (ruudk) [inline src/Application/PHPStanRunner.php:18]: Should be imported

File: src/Config/ConfigBuilder.php [hunk 11]
```diff
@@ -20,6 +22,10 @@ final class ConfigBuilder
     private ?string $phpstanConfigurationFile = null;
     private null | false | string $phpstanMemoryLimit = null;
     private ?string $twigEnvironmentLoader = null;
+    /**
+     * @var list<class-string<TemplateContextCollector<Node>>>
+     */
+    private array $twigContextCollectors = [];
```
Comment (ruudk) [inline src/Config/ConfigBuilder.php:28]:
```suggestion

    /**
     * @var list<class-string<TemplateContextCollector<Node>>>
     */
    private array $twigContextCollectors = [];
```

**Comment by raffaelecarelle** [comment]
Hello @ruudk, for the test case how do you suggest to proceed? 

Add a ConfigurationTest? Or PhpStanRunnerTest? 

Thank you

**Comment by ruudk** [comment]
An end2end test in Renderpoints. One that sets up a custom Collector and then verifies it works.

I pushed some changes to main, could you rebase it and fix the remainder CS issue? 

If you have [castor installed](https://castor.jolicode.com/getting-started/installation/#installation), you can use `castor fmt` and or `castor qa` to run all QA tasks.

**Comment by raffaelecarelle** [comment]
> An end2end test in Renderpoints. One that sets up a custom Collector and then verifies it works.
> 
> I pushed some changes to main, could you rebase it and fix the remainder CS issue?
> 
> If you have [castor installed](https://castor.jolicode.com/getting-started/installation/#installation), you can use `castor fmt` and or `castor qa` to run all QA tasks.

cool, I didn't know castor tool!

I added a test case. Take a look. Thank you

**Comment by ruudk** [comment]
Great one, thank you 🎉

**Pull request merged**